handout/testing/Unit Testing Components.md

```diff
@@ -0,0 +1,240 @@
+    beforeEach(() => {
+        this.app = new MessageComponent();
+    });
+
+    it('should set new message', () => {
+        this.app.setMessage("Testing")
+        expect(this.app.message).toBe("Testing");
+    });
+
+	it('should clear message', () => {
+       this.app.clearMessage("");
+	   expect(this.app.message).toBe("");
+    });
+});
+```
+
+We have created two tests, one for `setMessage`, and the other for `clearMessage`. In order to call those functions on the `MessageComponent` class we have to initiate it, and we do this with the `beforeEach` function. The `beforeEach` function is called before each test is performed, it is where all initiate required for testing takes place. Once our `MessageComponent` object is created we can call `setMessage` and `clearMessage` and anaylize the results of those actions. We formulate an expected result, and then test to see if the result we were expecting came to be. Here we are testing whether or not the message we tried to set modified the `MessageComponent` property `message` to the value we intended. If it did, then the test was sucessful and our `MessageComponent` works as expected. 
+
+## Injecting Dependencies and DOM Changes
+
+In the previous example the class we were testing `MessageComponent` did not have any injected dependencies. In Angular 2, components will often rely on services and other components in order to function, and these will be injected into the constructor of the class. When testing these components we have to inject the dependencies ourselves. Since this is an Angular specific routine, there are no pure Jasmine functions used to do this. Angular provides a list of functions in `angular2/testing` that allow us to to fully test our components. Lets take a look at a basic component:
+
+*quote.component.ts*
+
+``` typescript
+import {QuoteService} from './quote.service';
+import {Component} from 'angular2/core';
+
+@Component({
+	selector: 'my-quote',
+    template: '<h3>Random Quote</h3> <div>{{quote}}</div>'
+});
+
+export class QuoteComponent {
+  quote: string;
+  
+  constructor (private quoteService: QuoteService){}
+  
+  getQuote() {
+  	this.quoteService.getQuote().then((quote) => {
+  		this.quote = quote;
+	});
+   }
+}
+```
+
+This component relies on the `QuoteService` to get a random quote, which it will then display. The class is pretty simple, it only has the `getQuote` function that will modify the DOM, therefore it will be are main area of focus in testing.
+
+In order to test this component we need initiate the class `QuoteComponent`, the Angular testing library offers a way to inject any dependencies this class relies on, as well as a `TestComponentBuilder` object which will create the component for us and return a *component fixture* that we can perform tests on. 
+
+*quote.spec.ts*
+
+``` typescript
+class MockQuoteService {
+  public quote: "Test quote";
+  
+  getQuote() {
+  	return this.quote;
+  }
+}
```

Comment:
Is not a problem that `MockQuoteService` is returning a plain value when the original service is returning a promise?